test/UpdaterAndFilterEngineCreation.cpp

```diff
@@ -0,0 +1,158 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Adblock Plus.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <string>
+#include <memory>
+#include <array>
+
+#include <AdblockPlus/AsyncExecutor.h>
+
+#include "BaseJsTest.h"
+
+using namespace AdblockPlus;
+
+namespace
+{
+  class UpdaterAndFilterEngineCreationTest : public BaseJsTest
+  {
+  protected:
+
+    static const size_t COUNT = 100;
+    const std::string PROP_NAME = "patternsbackupinterval";
+    std::array<Updater*, COUNT> updaterAddrArray;
+    std::array<FilterEngine*, COUNT> filterAddrArray;
+    DelayedWebRequest::SharedTasks webRequestTasks;
+    DelayedTimer::SharedTasks timerTasks;
+
+    void SetUp()
+    {
+      LazyFileSystem* fileSystem;
+      ThrowingPlatformCreationParameters platformParams;
+      platformParams.logSystem.reset(new LazyLogSystem());
+      platformParams.timer = DelayedTimer::New(timerTasks);
+      platformParams.fileSystem.reset(fileSystem = new LazyFileSystem());
+      platformParams.webRequest = DelayedWebRequest::New(webRequestTasks);
+      platform.reset(new Platform(std::move(platformParams)));
+      std::uninitialized_fill(updaterAddrArray.begin(), updaterAddrArray.end(), nullptr);
+      std::uninitialized_fill(filterAddrArray.begin(), filterAddrArray.end(), nullptr);
+    }
+
+    void CallGetUpdaterSimultaneously()
+    {
+      CallGetSimultaneously(true, false);
+    }
+
+    void CallGetFilterEngineSimultaneously()
+    {
+      CallGetSimultaneously(false, true);
+    }
+
+    void CallGetUpdaterAndGetFilterEngineSimultaneously()
+    {
+      CallGetSimultaneously(true, true);
+    }
+
+  private:
+
+    void CallGetSimultaneously(bool isUpdater, bool isFilterEngine)
+    {
+      AsyncExecutor asyncExecutor;
+      for (size_t idx = 0; idx < COUNT; ++idx)
+        asyncExecutor.Dispatch(([this, idx, isUpdater, isFilterEngine]() -> void {
+          if (isUpdater && isFilterEngine)
+          {
+            // some randomization in order of calling gets
+            if (idx % 2)
+            {
+              updaterAddrArray[idx] = &(platform->GetUpdater());
+              filterAddrArray[idx] = &(platform->GetFilterEngine());
+            }
+            else
+            {
+              filterAddrArray[idx] = &(platform->GetFilterEngine());
+              updaterAddrArray[idx] = &(platform->GetUpdater());
+            }
+          }
+          else if (isUpdater)
+            updaterAddrArray[idx] = &(platform->GetUpdater());
+          else if (isFilterEngine)
+            filterAddrArray[idx] = &(platform->GetFilterEngine());
+        }));
+    }
+  };
+}
+
+TEST_F(UpdaterAndFilterEngineCreationTest, TestFilterEngineSingleInstance)
+{
+    CallGetFilterEngineSimultaneously();
+    FilterEngine* filterEngineAddr = filterAddrArray[0];
+    ASSERT_NE(filterEngineAddr, nullptr);
```

Comment:
Sorry for not noticing it before. Even if it's a nullptr, we still can continue the execution of the test. It's a pretty subtle and perhaps even subjective thing, but could you please change all these ASSERT_* to EXPECT_*. It can seem inconsistent from other tests but we are gradually applying [this approach](https://github.com/google/googletest/blob/master/googletest/docs/primer.md#assertions).

> Usually EXPECT_* are preferred, as they allow more than one failure to be reported in a test. However, you should use ASSERT_* if it doesn't make sense to continue when the assertion in question fails.

Therefore we usually use ASSERT_ in such cases when we know that the normal execution should not be possible, e.g. of a seg-fault because of dereferencing of a nullptr (in general case it can of course work without a seg-fault).

BTW, normally the order of arguments is that the expected value is going first and the actual value is the second one. Could you please change that order too if it does not cause compilation errors?

Reply:
OK, I know this appoach - I'll change those nullptr checks from ASSERT to EXPECT and will update the order, I'll push new commit in couple of minutes.